File: src/main/kotlin/java/_04_stratified_design/megamart/stratified/dom/DOM.java
package java._04_stratified_design.megamart.stratified.dom;

import java._04_stratified_design.megamart.stratified.BusinessRules;
import java._04_stratified_design.megamart.stratified.Cart;
import java._04_stratified_design.megamart.stratified.CartItem;
import java._04_stratified_design.megamart.stratified.CartRules;

import java.util.List;

public class DOM {

    public static void addItemToCart(List<CartItem> cart, String name, Double price) {
        cart = Cart.addItem(cart, new CartItem(name, price));    // Relegate addition of cart items to Cart layer
        var total = Cart.calcTotal(cart);                        // Cart layer handles total calculation
        setCartTotal(total);                                     // Setting the total in the DOM is part of this layer
        updateShippingIcons(cart);                               // ...as is updating the shipping icons
        updateTaxDom(total);                                     // ...as is updating the tax in the DOM
    }

    public static List<BuyButton> getBuyButtons() {
        return List.of(
                new BuyButton(new CartItem("Shoes", 19.99)),
                new BuyButton(new CartItem("Shirt", 9.99)),
                new BuyButton(new CartItem("Pants", 14.99)),
                new BuyButton(new CartItem("Socks", 4.99))
        );
    }

    public static void setCartTotal(Double total) {
        System.out.println("Cart total: " + total);
    }

    public static void setTax(Double tax) {
        System.out.println("Tax: " + tax);
    }

    // LAYER: DOM operations
    public static void updateShippingIcons(List<CartItem> cart) {

        var buyButtons = BuyButton.deepCopy(getBuyButtons());  // The deep copying code is part of the BuyButton class now, a better 'fit'

        for (var button : buyButtons) {
            // The 'getsFreeShipping' function is now part of the CartRules layer
            // The 'addItem' function is in the lower-level Cart layer
            if (CartRules.getsFreeShipping(Cart.addItem(cart, button.item()))) {
                button.showFreeShippingIcon();
            } else {
                button.hideFreeShippingIcon();
            }
        }
    }

    public static void updateTaxDom(double shoppingCartTotal) {
        // Calculation of a tax amount is now part of the BusinessRules layer
        setTax(BusinessRules.calcTax(shoppingCartTotal));
    }


}

File: src/main/kotlin/_04_stratified_design/megamart/stratified/dom/Dom.kt
package _04_stratified_design.megamart.stratified.dom

import _04_stratified_design.megamart.stratified.cart.*
import _04_stratified_design.megamart.stratified.marketing.*
import _04_stratified_design.megamart.stratified.businessRules.*

fun addItemToCart(cart: List<CartItem>, name: String, price: Double): List<CartItem> {
    var updatedCart = addItem(cart, CartItem(name, price))

    refreshDom(updatedCart)

    return updatedCart
}

fun deleteHandler(cart: List<CartItem>, name: String) {
    val updatedCart = removeItemByName(cart, name)
    refreshDom(updatedCart)
}

fun refreshDom(cart: List<CartItem>) {
    val total = calcTotal(cart)
    setDOMCartTotal(total)
    updateDOMShippingIcons(cart)
    updateDOMTax(total)
}

fun updateDOMShippingIcons(cart: List<CartItem>) {
    val buyButtons = getDOMBuyButtons() // Accessing the DOM is an action
    for (button in buyButtons) {
        if (getsFreeShipping(addItem(cart, button.item))) {
            println("Free shipping icon shown on ${button.item.name}")
        } else {
            println("Free shipping icon hidden on ${button.item.name}")
        }
    }
}

fun setDOMCartTotal(total: Double) {
    // Update the total in the DOM (an action)
    // e.g. document.getElementById("cart-total").innerText = total
    // To keep things simple here, we'll just print the change
    println("Updated cart total to $total")
}

fun updateDOMTax(total: Double) {
    // val tax = shoppingCartTotal * 0.13  // Reading a global variable is an action
    // By passing in the total as an explicit input, we can now calculate the tax without relying on a global variable
    println("Updated tax to ${calcTax(total)}")
}

data class BuyButton(val item: CartItem)
fun getDOMBuyButtons() =
// NOTE: This is just a simulation of something like
// document.getElementsByClassName("buy-button")

    // To keep things simple here, we just simulate with a list of 3 buttons
    listOf(
        BuyButton(CartItem("Item 1", 10.0)),
        BuyButton(CartItem("Item 2", 15.0)),
        BuyButton(CartItem("Item 3", 5.0))
    )